src/server/sendLayoutHTTPResponse.js

```diff
@@ -215,9 +224,16 @@ function serializeApplication({
     `Application ${appName}`
   );
 
+  // If there are assets streamed then add them to the assetsStream
+  if (appStream.assets) {
```

Comment:
To facilitate full control over performance, i think we should allow for the `assets` and `content` properties to be promises that resolve with the stream/string. That way you can start providing the assets before you even have the content. The implementation here currently does not allow for that, and should be changed.

Unfortunately, doing this makes things a lot more complicated. The `renderResultToStream` function is shared between fragments and applications and we probably can't do that anymore. The problem here is that we need to add the stream to bodyStream synchronously, but the stream might not have content written to it asynchronously. I think the best approach might be to get rid of the `renderResultToStream` function entirely, and replace with a less comprehensive `valueToStream` function:

```js
function valueToStream(value, name) {
  let stream

  if (value && typeof value.then === "function") {
    const promise = value;
    stream = merge2();
    promise.then(
      (result) => {
        stream.add(
          typeof result === "string" ? stringStream(result) : result
        );
      },
      (err) => {
        stream.add(renderError(name, err));
      }
    );
  } else if (typeof value === "string") {
    stream = stringStream(value);
  } else {
    stream = value
  }

  return stream;
}
```

Then the renderApplication and renderFragment functions would have to change accordingly:

```js
function serializeFragment() {
  // this isn't the full implementation of serializeFragment, just the stream part.


  let fragmentStream;
  try {
    fragmentStream = valueToStream(renderOptions.renderFragment(name));
  } catch (err) {
    fragmentStream = renderError(`Fragment ${name}`, err);
  }
}

function serializeApplication() {
  // this isn't the full implementation of serializeApplication, just the stream part.

  let renderResult, contentStream, assetStream;
  try {
    renderResult = renderOptions.renderApplication(...);

    if (typeof renderResult.then === 'function') {
      contentStream = merge2();
      assetStream = merge2();
      renderResult.then(
        (value) => {
          const streams = valueToAppStreams(name, value);
          contentStream.add(streams.contentStream);
          assetStream.add(streams.assetStream);
        },
        (err) => {
          contentStream.add(renderError(name, err));
          assetStream.add(renderError(name, err));
        }
      )
    } else {
      const streams = valueToAppStreams(name, renderResult);
      contentStream = streams.contentStream;
      assetStream = streams.assetStream;
    }
  } catch (err) {
    contentStream = renderError(name, err);
    assetStream = renderError(name, err);
  }

  assetStream.add(assetStream);
  bodyStream.add(contentStream);
}

function valueToAppStreams(name, value) {
  let contentStream, assetStream
  
  if (value) {
    contentStream = valueToStream(name, value.content || value);
    assetStream = valueToStream(name, value.assets || value);
  } else {
    contentStream = renderError(name, Error(`returned nothing`));
    assetStream = renderError(name, Error(`returned nothing`));
  }

  return { contentStream, assetStream }
}
```

Reply:
Added this with two changes
1) valueToAppStreams - The assetStream will have either value.assets or an empty string 
2) serializeLayoutData also needed to be updated with the removal of `renderResultToStream` with `valueToStream`